networkx/algorithms/shortest_paths/unweighted.py

```diff
@@ -72,25 +72,24 @@ def _single_shortest_path_length(adj, firstlevel, cutoff):
         cutoff : int or float
             level at which we stop the process
     """
-    seen = {}  # level (number of hops) when seen in BFS
-    level = 0  # the current level
-    nextlevel = set(firstlevel)  # set of nodes to check at next level
+    seen = set(firstlevel)
+    nextlevel = list(firstlevel)
+    level = 0
     n = len(adj)
-    while nextlevel and cutoff >= level:
-        thislevel = nextlevel  # advance to next level
-        nextlevel = set()  # and start a new set (fringe)
-        found = []
-        for v in thislevel:
-            if v not in seen:
-                seen[v] = level  # set the level of vertex v
-                found.append(v)
-                yield (v, level)
-        if len(seen) == n:
-            return
-        for v in found:
-            nextlevel.update(adj[v])
+    for v in nextlevel:
+        yield (v, level)
+    while nextlevel and cutoff > level:
         level += 1
-    del seen
+        thislevel = nextlevel
+        nextlevel = []
+        for v in thislevel:
+            for w in adj[v]:
+                if w not in seen:
+                    seen.add(w)
+                    nextlevel.append(w)
+                    yield (w, level)
+            if len(seen) == n:
+                return
```

Comment:
I was thinking about the `del seen` code that has been removed in this PR.  I think we need it to avoid large memory hoarding by the (potentially many) exhausted iterator/generator.  Granted this is only a problem if the user code is calling this many times and not releasing the references to the generator. But I can imagine that happening, and that was the concern driving having `del seen` in the current code.

The potential problem is that a large set `seen` is created in the generator. If we don't delete that set before `return`, the generator stops yielding results, but it can avoid "garbage collection" if any references to the generator in the user code aren't removed. Usually keeping exhausted iterators is not a memory problem because they are small. But in this can `seen` could be fairly large.  

What do you think?  The fix would be:
```suggestion
            if len(seen) == n:
                del seen
                return
    del seen
```

Reply:
mmh I don't think it's useful to have that because when you hit return in a generator it raises StopIteration (see https://stackoverflow.com/questions/16780002/return-in-generator-together-with-yield) which starts the garbage collection.

As an example I made:

```
def gen():
    a = [i for i in range(10**6)]
    for x in a:
        if x != 1:
            yield x
        else:
            return

k = gen()
next(k) # memory usage increases
next(k) # memory is freed
```

Am I misunderstanding your point?

Reply:
Unless something has changed in python since I last looked at this (a long time ago), the memory is not freed when an iterator is exhausted. In your example, the memory would not be freed until the variable `k` releases its reference to the iterator object. For example if you did `del k` then it would release the memory used by the list `a`. Certainly the variable `k` still exists after the StopIteration and you can call `next(k)` again even after the StopIteration has been raised.  But maybe Python is smart enough now to delete the local namespace of the instance when the StopIteration is raised. But I don't think so. That could cause problems with generators that get restarted.

Reply:
Or are there any subtlety I don't see? Can you point me out a concrete example if this is the case? :-)

Reply:
You are right!   I think this must be an out-of-date understanding that I had about generators. It looks like StopIteration releases all the refcounts within the generator.

Thanks!  I learned something ---   makes me wonder if there are other `del something` in the code base. I'll search for them... :)

Reply:
glad to know it! :-)